### ISSUE TITLE
The response for ActionDispatch::RemoteIp::IpSpoofAttackError is not 4xx, but 500

### ISSUE THREAD

**Issue by willnet** [rails/rails#46003]
On IpSpoofAttackError raising, I want rails to return 4xx since it is a client-side problem.
Since the ActionDispatch::RemoteIp middleware is located before ActionDispatch::ShowExceptions, We cannot define the behavior when the exception raises with `config.action_dispatch.rescue_responses`.

ActionDispatch::HostAuthorization is also located before ActionDispatch::ShowExceptions, but it has its own Rack application to return 403 on exceptions. Maybe ActionDispatch::RemoteIp should have its own Rack appilcation as well.

### Steps to reproduce

```shell
rails new sampleapp
cd sampleapp
rails s -d
curl localhost:3000 -H "X-Forwarded-For: 1.2.3.4" -H "Client-Ip: 5.6.7.8" -I
```

### Expected behavior

response 4xx

### Actual behavior

```
HTTP/1.1 500 Internal Server Error
Content-Length: 4602
```

### System configuration
**Rails version**: 7.0.4

**Ruby version**: 3.1.2

**Comment by casperisfine** [comment]
Hum, it's always hard to know why an order was picked, but I wonder if here it's not because of the logger:

```ruby
          middleware.use ::ActionDispatch::RequestId, header: config.action_dispatch.request_id_header
          middleware.use ::ActionDispatch::RemoteIp, config.action_dispatch.ip_spoofing_check, config.action_dispatch.trusted_proxies

          middleware.use ::Rails::Rack::Logger, config.log_tags
          middleware.use ::ActionDispatch::ShowExceptions, show_exceptions_app
```

e.g. on the way in `RemoteIP` is before the logger, so you log the right IP, then on the way back, if an uncaught exception was turned into an error page, you log the correct status code.

So it's not possible to change that order without breaking the above :/, but then yes it's totally incorrect for `ActionDispatch::RemoteIp` to raise.

These middleware ordering issues are always headache inducing -_-.

@rafaelfranca any idea on this one?

**Comment by matthewd** [comment]
[The corresponding test](https://github.com/rails/rails/blob/b1c5ae394782e3a766c99012d9ad87476d095544/railties/test/application/middleware_test.rb#L119) is a bit more communicative about the ordering, though from a glance, not this particular pair.

I don't think the logger technically ever sees the response status. I'd be somewhat concerned if there were a risk that a different ordering could result in [even failed] requests being logged, but in practice that doesn't seem like a concern either.

In practice I suspect it would be safe to swap Logger and ShowExceptions... but that doesn't actually help us: just like Logger, it's important that ShowExceptions sees the post-RequestId and post-RemoteIp request state.

**Comment by casperisfine** [comment]
Agreed.

So I think the only solution is not to raise but to return a 4xx. 

Looking at the spec I think `400 Bad Request` is the one for this kind of scenarios:

> The server cannot or will not process the request due to an apparent client error (e.g., malformed request syntax, size too large, invalid request message framing, or **deceptive request routing**).

https://www.rfc-editor.org/rfc/rfc7231#section-6.5.1

**Comment by rafaelfranca** [comment]
Yeah. I think we should not be raising. There is no reason for users to have to configure how this middleware will respond. 400 makes sense to me.

**Comment by casperisfine** [comment]
Hum, I just realized that even returning a 400 isn't that simple when you are before `ShowException`.

I mean it's easy to `return [400, ...]`, but wouldn't use expect this to go through the exception_app so it's customizable?

**Comment by matthewd** [comment]
It also seems not-ideal to me for this middleware to be responsible for composing and returning a full HTTP response... an exception intuitively feels like a better fit for the level of abstraction it's interested in (despite the obvious realities of it being a Rack middleware).

Perhaps we could double up ShowExceptions, injecting another copy super early specifically to handle exceptions that occur in the space above the main one? I'm nervous about exposing a user's exception_app to requests in that state (e.g. in this case, having potentially misleading client IP info[^1]), but maybe that's just the least bad option.

[^1]: I'm specifically wary that this is a reasonable thing to actually _use_ in an exception app: for example, one might return a more detailed [and potentially revealing] error response to an apparently-internal client.

**Comment by rafaelfranca** [comment]
Do this response really needs to be customizable? I don't see a case that people want to show an attacker a custom response.

**Comment by casperisfine** [comment]
> What if we push this middleware down them? Is there any reason why it is before `ShowException`?

I think it's because of the logger: https://github.com/rails/rails/issues/46003#issuecomment-1245340450
It's a big chicken and egg problem.

>  don't see a case that people want to show an attacker a custom response.

Well, we assume it's an attacker, it could be a misconfigured proxy or something like that. But yes, that probably wouldn't be the end of the world.

**Comment by willnet** [comment]
What caused me to create this Issue was an error in our customer's environment. No attackers.

Therefore, it would be nice to be able to customize the response, but just returning 400 is enough.

**Comment by Auralcat** [comment]
I could reproduce the 500 response error in both the current `7-0-stable` branch and the `main` branch at 18acbe8948e04088a3116ed3eb5df05ce72f0b20

For reference, this is the middleware chain I got through `rails middleware` when reproducing the issue for both branches:


# 7-0-stable
```
use ActionDispatch::HostAuthorization
use Rack::Sendfile
use ActionDispatch::Static
use ActionDispatch::Executor
use ActionDispatch::ServerTiming
use ActiveSupport::Cache::Strategy::LocalCache::Middleware
use Rack::Runtime
use Rack::MethodOverride
use ActionDispatch::RequestId
use ActionDispatch::RemoteIp
use Sprockets::Rails::QuietAssets
use Rails::Rack::Logger
use ActionDispatch::ShowExceptions
use WebConsole::Middleware
use ActionDispatch::DebugExceptions
use ActionDispatch::ActionableExceptions
use ActionDispatch::Reloader
use ActionDispatch::Callbacks
use ActiveRecord::Migration::CheckPending
use ActionDispatch::Cookies
use ActionDispatch::Session::CookieStore
use ActionDispatch::Flash
use ActionDispatch::ContentSecurityPolicy::Middleware
use ActionDispatch::PermissionsPolicy::Middleware
use Rack::Head
use Rack::ConditionalGet
use Rack::ETag
use Rack::TempfileReaper
run IpSpoofIssue::Application.routes
```

# `main` branch

```
use ActionDispatch::HostAuthorization
use Rack::Sendfile
use ActionDispatch::Static
use ActionDispatch::Executor
use ActionDispatch::ServerTiming
use ActiveSupport::Cache::Strategy::LocalCache::Middleware
use Rack::Runtime
use Rack::MethodOverride
use ActionDispatch::RequestId
use ActionDispatch::RemoteIp
use Sprockets::Rails::QuietAssets
use Rails::Rack::Logger
use ActionDispatch::ShowExceptions
use WebConsole::Middleware
use ActionDispatch::DebugExceptions
use ActionDispatch::ActionableExceptions
use ActionDispatch::Reloader
use ActionDispatch::Callbacks
use ActiveRecord::Migration::CheckPending
use ActionDispatch::Cookies
use ActionDispatch::Session::CookieStore
use ActionDispatch::Flash
use ActionDispatch::ContentSecurityPolicy::Middleware
use ActionDispatch::PermissionsPolicy::Middleware
use Rack::Head
use Rack::ConditionalGet
use Rack::ETag
use Rack::TempfileReaper
run IpSpoofIssue::Application.routes
```
